File: OCaml/d01/ex06/iter.ml
let iter f x n =
  if n < 0 then (-1) else
  let rec loop = function
      (y, 0) -> y
    | (y, m) -> loop((f y), (m - 1))
  in loop(x, n)

let () =
  print_int(iter (fun x -> x * x) 2 4); print_char '\n';
  print_int(iter (fun x -> x * 2) 2 4); print_char '\n'
  

File: cpp/d07/ex01/iter.cpp
#include <iostream>

template <typename T>
void iter(T a[], unsigned len, void (*fn)(T e)) {
	for (unsigned i = 0; i < len; ++i) fn(a[i]);
}

void print_int(int i) {
	std::cout << i << std::endl;
}

int main() {
	int a[] = { 5, 4, 3, 2, 1 };

	iter(a, 5, print_int);

	return 0;
}
